src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java

```diff
@@ -0,0 +1,782 @@
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.internal;
+
+import static com.google.common.truth.Truth.assertWithMessage;
+
+import java.io.File;
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import javax.xml.parsers.ParserConfigurationException;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.w3c.dom.Document;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+import com.google.common.collect.ImmutableMap;
+import com.puppycrawl.tools.checkstyle.AbstractModuleTestSupport;
+import com.puppycrawl.tools.checkstyle.Checker;
+import com.puppycrawl.tools.checkstyle.DefaultConfiguration;
+import com.puppycrawl.tools.checkstyle.ModuleFactory;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.Scope;
+import com.puppycrawl.tools.checkstyle.api.SeverityLevel;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.checks.LineSeparatorOption;
+import com.puppycrawl.tools.checkstyle.checks.annotation.AnnotationUseStyleCheck;
+import com.puppycrawl.tools.checkstyle.checks.blocks.BlockOption;
+import com.puppycrawl.tools.checkstyle.checks.blocks.LeftCurlyOption;
+import com.puppycrawl.tools.checkstyle.checks.blocks.RightCurlyOption;
+import com.puppycrawl.tools.checkstyle.checks.imports.ImportOrderOption;
+import com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocContentLocationOption;
+import com.puppycrawl.tools.checkstyle.checks.naming.AccessModifierOption;
+import com.puppycrawl.tools.checkstyle.checks.whitespace.PadOption;
+import com.puppycrawl.tools.checkstyle.checks.whitespace.WrapOption;
+import com.puppycrawl.tools.checkstyle.internal.utils.TestUtil;
+import com.puppycrawl.tools.checkstyle.internal.utils.XdocUtil;
+import com.puppycrawl.tools.checkstyle.internal.utils.XmlUtil;
+import com.puppycrawl.tools.checkstyle.utils.CheckUtil;
+import com.puppycrawl.tools.checkstyle.utils.JavadocUtil;
+import com.puppycrawl.tools.checkstyle.utils.ScopeUtil;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+public class XdocsJavaDocsTest extends AbstractModuleTestSupport {
```

Comment:
This is not a new class. Not sure why it wasn't disabled instead of deleting.

Can you provide a diff to compare what it was before and what it is now?

Reply:
Here you go. 2b849e426b06eb9b1777476de4c1657c71a373ba is the commit before the test was removed so this was the latest state of it.
```bash
stoyan @ legion: ~/open-source/checkstyle                                                                                                                                      pull/13361
$ git --no-pager diff 2b849e426b06eb9b1777476de4c1657c71a373ba HEAD src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java

```
```diff
diff --git a/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java b/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java
index cc3964b76..c40b48aba 100644
--- a/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java
@@ -31,6 +31,7 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.Set;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
@@ -120,6 +121,8 @@ public class XdocsJavaDocsTest extends AbstractModuleTestSupport {
 
     private static String checkName;
 
+    private static Set<Path> xdocsFilePaths;
+
     @Override
     protected String getPackageLocation() {
         return "com.puppycrawl.tools.checkstyle.internal";
@@ -130,6 +133,7 @@ public class XdocsJavaDocsTest extends AbstractModuleTestSupport {
         final DefaultConfiguration checkConfig = new DefaultConfiguration(
                 JavaDocCapture.class.getName());
         checker = createChecker(checkConfig);
+        xdocsFilePaths = XdocUtil.getAllXdocFilePaths();
     }
 
     /**
@@ -147,7 +151,8 @@ public class XdocsJavaDocsTest extends AbstractModuleTestSupport {
             final File file = path.toFile();
             final String fileName = file.getName();
 
-            if ("config_system_properties.xml".equals(fileName)) {
+            if ("config_system_properties.xml".equals(fileName)
+                    || "index.xml".equals(fileName)) {
                 continue;
             }
 
@@ -518,22 +523,13 @@ public class XdocsJavaDocsTest extends AbstractModuleTestSupport {
             final String attrValue;
 
             if ("a".equals(nodeName) && "href".equals(attrName)) {
-                String value = attribute.getNodeValue();
+                final String value = attribute.getNodeValue();
 
                 assertWithMessage("links starting with '#' aren't supported: " + value)
                     .that(value.charAt(0))
                     .isNotEqualTo('#');
 
-                if (value.contains("://")) {
-                    attrValue = value;
-                }
-                else {
-                    if (value.charAt(0) == '/') {
-                        value = value.substring(1);
-                    }
-
-                    attrValue = "https://checkstyle.org/" + value;
-                }
+                attrValue = getLinkValue(value);
             }
             else {
                 attrValue = attribute.getNodeValue();
@@ -548,6 +544,57 @@ public class XdocsJavaDocsTest extends AbstractModuleTestSupport {
         return result.toString();
     }
 
+    private static String getLinkValue(String initialValue) {
+        String value = initialValue;
+        final String attrValue;
+        if (value.contains("://")) {
+            attrValue = value;
+        }
+        else {
+            if (value.charAt(0) == '/') {
+                value = value.substring(1);
+            }
+
+            if (value.contains("../apidocs/")
+                    || value.contains("../files/")
+                    || value.contains("../styleguides")
+                    || value.contains("index.html")) {
+                value = value.replaceAll("\\.\\./", "");
+            }
+            else if (!value.contains("dtds/")) {
+                final int lastIndexOfSlash = value.lastIndexOf('/');
+                final String fileNameWithoutExtension;
+                if (lastIndexOfSlash == -1) {
+                    fileNameWithoutExtension = value.substring(0, value.indexOf(".html"));
+                }
+                else {
+                    fileNameWithoutExtension = value
+                            .substring(lastIndexOfSlash + 1, value.indexOf(".html"));
+                }
+                final Optional<Path> xdocsFilePath = xdocsFilePaths
+                        .stream()
+                        .filter(path -> {
+                            return path
+                                    .toString()
+                                    .matches(".*[\\\\/]" + fileNameWithoutExtension + "\\.xml.*");
+                        })
+                        .findFirst();
+                if (xdocsFilePath.isPresent()) {
+                    final String filePathFromRoot = xdocsFilePath.get().toString()
+                            .replaceAll("src[\\\\/]xdocs[\\\\/]", "");
+                    value = filePathFromRoot
+                            .substring(0, filePathFromRoot.lastIndexOf(".xml"))
+                            + ".html"
+                            + value.substring(value.indexOf(".html") + 5);
+                    value = value.replaceAll("\\\\", "/");
+                }
+            }
+
+            attrValue = "https://checkstyle.org/" + value;
+        }
+        return attrValue;
+    }
+
     public static class JavaDocCapture extends AbstractCheck {
         private static final Pattern SETTER_PATTERN = Pattern.compile("^set[A-Z].*");
 

```

Reply:
Why is `getLinkValue` so complicated over it's original. I am definately not understand the `dtd` pathing. We have links to something under DTDs?

Reply:
Something just seems wrong with this method unless there is more I am missing. For `../` you should be able to identify the folder you are in and determine the full path from it without having to look up through `xdocsFilePaths` or some of the other weirdness going on here.

For example, if my path is `src\xdocs\checks\design\finalclass.xml`,
`` = `checkstyle.org/checks/design/`
`../` = `checkstyle.org/checks/`
`../../` = `checkstyle.org/`

I don't need to know the full list of files involved, I just need to know where the base is.

Reply:
This looks a lot cleaner even if I don't fully know what `resolve` and `normalize` are doing.

Reply:
```java
value; // ../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CTOR_DEF
value = currentXdocPath // src/xdocs/checks/whitespace/methodparampad.xml
        .getParent() // src/xdocs/checks/whitespace
        .resolve(Paths.get(value)) // src/xdocs/checks/whitespace/../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CTOR_DEF
        .normalize() // src/xdocs/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CTOR_DEF
        .toString()
        .replaceAll("src[\\\\/]xdocs[\\\\/]", "") // apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#CTOR_DEF
        .replaceAll("\\\\", "/");
```